### DISCUSSION TITLE
Can OIDC client handle revoked tokens?

### DISCUSSION THREAD

**Discussion by IvanPuntev** [quarkusio/quarkus#43904] [Q&A]
Hi @sberyozkin. Can you tell me if the oidc client can handle revoked jwt tokens? For example if the client already got token but the backend decides to revoke it soon after and I try to call this backend with the old token will the library handle this situation and get new token?

[a comment by an automated account is not shown]

**Comment by sberyozkin** [comment]
Hi @IvanPuntev, 

In this case we have 401 from the target, and given that OIDC request client filter itself does not see this 401, I suggested, [here](https://github.com/quarkusio/quarkus/issues/43799#issuecomment-2404714929) 2 options:
* Have a custom response filter coordinating with the custom OidcClientRequest filter extension, it can really work though for a single OIDC client setup case, as the response filter will have no idea which OIDC client token caused the current 401. If it is a single client only, then the response filter can set a `new token required` flag for the request filter to react upon it during the retry
* Or, use `OidcClient` directly as linked in that comment.

I don't see how else OidcClient can know that what is a valid token from its own perspective, has already been revoked.
Does it help ?

**Reply by IvanPuntev** [reply]
Thanks for the help @sberyozkin. But what exactly is the situation with the library that it won't handle the 401 response and retry to get new token by default? Also what is the behaviour when no exp property is returned from the token response? How the library knows it needs to get new token?

**Reply by sberyozkin** [reply]
@IvanPuntev 

> But what exactly is the situation with the library that it won't handle the 401 response and retry to get new token by default? 

OIDC client filter supports a case where users prefer not to deal with OidcClient directly.

`OidcClient` running in a client request filter does not see responses to its requests, they are coming in the client response filter. Without a hint (as suggested above), when it sees the token which it still sees as valid, it can only reuse it.
In theory, we could've shipped the client response filter which would set a `new token required` flag in case of 401, which may work for the single OIDC client case, but then it would impact all the applications which use more than one OIDC client, because at the client response filter level it is impossible to figure out which specific OIDC client acquired the token which now caused 401.

So, as I suggested, one can register a custom client response filter to set a `new token required` flag in case of 401, or take complete control with using `OidcClient` directly.

>Also what is the behaviour when no exp property is returned from the token response? How the library knows it needs to get new token?

Good question :-). It does not know it needs a new token if some totally non-compliant OAuth2 server chooses not to tell to the consumers how long the token is valid for - it is valid forever.  Again, users have a choice and can use OidcClient directly and decide what to do. If they use a filter, either a `new token required` flag in case of 401 is expected, or, from 3.16, they can restrict its age with #43417.

Does it clarify things ? 

Like I said, we could've shipped a custom response filter, which would only set  `new token required`  on some shared property holder in case of 401 - but would force all OIDC client acquire new tokens...

What do you think ?

**Reply by IvanPuntev** [reply]
This clarifies it, thanks. But what stops the client response filter to figure out which specific OIDC client acquired the token which caused 401? Is there a way we can rework it and this is handled out of the box per oidc client?

**Reply by sberyozkin** [reply]
@IvanPuntev 

>But what stops the client response filter to figure out which specific OIDC client acquired the token which caused 401

Let's say we have 2 clients accessed concurrently. And all the client response filter can see is 401. How to link this 401 to a specific OIDC client ?

**Reply by sberyozkin** [reply]
@IvanPuntev Well, one can imagine that that the request filter sets an OIDC client name in the request context and the client response filter can check a request context and update that map where client names are keys that the new token is required. It will require though to have a static map in the abstract filter code, and an external retry.

Would you like to experiment and have OIDC Client reactive filter implement client response filter too ?

**Reply by IvanPuntev** [reply]
@sberyozkin I would like to try implementing this when I have some spare time. I would need some pointers where to start looking in the code to save some time tracing the flow.

**Reply by sberyozkin** [reply]
@IvanPuntev Can you open an enhancement request please ? I may give it a try as well. 

It will have to be optional anyway, there could be code in place, custom response filters which do some work related to 401. There could be a risk of the loop since 401 may be reported on non expired non revoked tokens.

However   may be we can do something useful with respect to supporting retries, thanks for driving the conversation

**Reply by IvanPuntev** [reply]
https://github.com/quarkusio/quarkus/issues/44037
